proposals/NNNN-modify-and-read-accessors.md

```diff
@@ -0,0 +1,621 @@
+        return x
+    }
+  }
+}
+```
+
+The problem is that `UniqueGet.property { get }` borrows the receiver and, executing like a normal function, transfers ownership of its result to the caller.
+
+This particular error could be addressed by marking the getter `consuming`:
+
+```swift
+struct UniqueString : ~Copyable {...}
+
+struct UniqueConsumingGet : ~Copyable {
+  var x: UniqueString
+
+  var property: UniqueString {
+    consuming get {
+        return x
+    }
+  }
+}
+```
+
+Now the getter takes ownership of the `UniqueConsumingGet`.
+That enables it to destructively extract `x` and transfer ownership of it to the caller.
+Here's how that looks in the caller:
+
+```swift
+let container = UniqueConsumingGet()
+let x = container.property // consumes container!
+// container is no longer valid
+```
+
+While for some things, this is desirable, for many typical uses of properties and subscripts, it is not.
+For example, if the container holds a number of noncopyable fields, it should be possible to inspect each in turn.
+Doing so wouldn't be possible if inspecting any one of them consumed the container.
+
+Similar to the mutating case, what's needed here is [an accessor](#design-read) which _borrows_ `self` and which _lends_ `x`--this time immutably--to the caller.
+
+## Proposed solution
+
+We propose two new accessor kinds:
+- `modify`, to enable mutating a value without first copying it
+- `read`, to enable inspecting a value without copying it.
+
+## Detailed design
+
+### Read<a name="design-read"/>
+
+[`UniqueGet`](#read-motivation) could allow its clients to inspect its field non-destructively with `read`:
+
+```swift
+struct UniqueString : ~Copyable {...}
+
+struct UniqueRead : ~Copyable {
+  var x: UniqueString
+
+  var property: UniqueString {
+    read {
+        yield x
+    }
+  }
+}
+```
+
+The `UniqueRead.property { read }` accessor is a "yield-once coroutine".
+When it is called, it borrows `self`.
+It runs until reaching a `yield` at which point it suspends, lending the yielded value back to the caller.
+Once the caller is finished with the value, it resumes the accessor.
+The accessor continues running where it left off, just after the `yield` where it suspended.
+
+If a `read` is provided, a `get` cannot also be provided.
+
+### Read as a protocol requirement
+
+Such accessors should be usable on values of generic and existential type.
+To indicate that a protocol provides immutable access to a property or subscript via a `read` coroutine,
+we propose allowing `read` to appear where `get` does today:
+
+```swift
+protocol Containing {
+  var property: UniqueString { read }
+}
+```
+
+If `read` is specified, `get` cannot also be specified.
+
+A `read` requirement can be witnessed by a stored property, a `read` accessor, a getter, or an unsafe addressor.
+
+### Get of noncopyable type as a protocol requirement
+
+Note that it not so easy to satisfy a `get` requirement whose type is noncopyable:
```

Comment:
Nit: Should read "Note that it _is_ not so easy to satisfy…"

Reply:
Thanks, fixed.